### PR TITLE
MF-922 - Add UpdateUser endpoint

### LINKED ISSUE #922: Add User Update endpoint
1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.
no
2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Update the User informations (metadata)
3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
must-have

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

Closes #922

### CODE REVIEW

File: users/api/http/transport.go [hunk 1]
```diff
@@ -56,6 +56,13 @@ func MakeHandler(svc users.Service, tracer opentracing.Tracer, l log.Logger) htt
 		opts...,
 	))
 
+	mux.Put("/users", kithttp.NewServer(
```
Comment (drasko) [inline users/api/http/transport.go:59]: Can we move this near to GET on `/users` (or vice-versa) - to increase readability.

File: users/tracing/users.go [hunk 2]
```diff
@@ -44,6 +44,14 @@ func (urm userRepositoryMiddleware) Save(ctx context.Context, user users.User) e
 	return urm.repo.Save(ctx, user)
 }
 
+func (urm userRepositoryMiddleware) Update(ctx context.Context, user users.User) error {
+	span := createSpan(ctx, urm.tracer, saveOp)
+	defer span.Finish()
+	ctx = opentracing.ContextWithSpan(ctx, span)
+
+	return urm.repo.Update(ctx, user)
```
Comment (drasko) [inline users/tracing/users.go]: Does this update whole user structure or just Metadata?

If it updates whole user structure, it means that User can change his own "constructive" information - which is contrary to the ide that users are created and managed by the admins - and can lead to security breaches.

@nmarcetic please review this one and advise.

**Review by drasko — CHANGES_REQUESTED** [review]

File: users/postgres/users.go [hunk 3]
```diff
@@ -43,8 +43,13 @@ func (ur userRepository) Save(ctx context.Context, user users.User) error {
 	return nil
 }
 
+type dbMetaUser struct {
+	Email    string     `db:"email"`
+	Metadata dbMetadata `db:"metadata"`
+}
```
Comment (anovakovic01) [inline users/postgres/users.go:58]: Why do we need this struct?

File: users/service.go [hunk 4]
```diff
@@ -142,6 +145,20 @@ func (svc usersService) UserInfo(ctx context.Context, token string) (User, error
 
 }
 
+func (svc usersService) UserUpdate(ctx context.Context, token string, u User) error {
```
Comment (anovakovic01) [inline users/service.go]: I recommend that you pass only metadata instead of the whole user.

File: users/api/http/requests.go [hunk 5]
```diff
@@ -28,6 +28,18 @@ func (req viewUserInfoReq) validate() error {
 	return nil
 }
 
+type updateUserReq struct {
+	token string
+	user  users.User
```
Comment (anovakovic01) [inline users/api/http/requests.go]: Please don't use user here (use metadata instead).

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: users/postgres/users.go (hunk 3 above)
Comment (manuio) [reply users/postgres/users.go:58]: I forgot to remove it, good catch!

File: users/api/http/transport.go (hunk 1 above)
Comment (drasko) [reply users/api/http/transport.go:59]: Indeed, sorry for this remark :)

File: users/tracing/users.go (hunk 2 above)
Comment (drasko) [reply users/tracing/users.go]: @nmarcetic please review and advise here so we can move further.

File: users/api/logging.go [hunk 6]
```diff
@@ -74,7 +74,20 @@ func (lm *loggingMiddleware) UserInfo(ctx context.Context, key string) (u users.
 		lm.logger.Info(fmt.Sprintf("%s without errors.", message))
 	}(time.Now())
 
-	return lm.svc.UserInfo(ctx, key)
+	return lm.svc.UserInfo(ctx, token)
+}
+
+func (lm *loggingMiddleware) UpdateMetadata(ctx context.Context, token string, metadata map[string]interface{}) (email string, err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method user_update for user %s took %s to complete", email, time.Since(begin))
```
Comment (anovakovic01) [inline users/api/logging.go]: Change `user_update` to `update_metadata`.

File: users/api/metrics.go [hunk 7]
```diff
@@ -56,13 +56,22 @@ func (ms *metricsMiddleware) Identify(key string) (string, error) {
...
 		ms.latency.With("method", "user_info").Observe(time.Since(begin).Seconds())
 	}(time.Now())
 
-	return ms.svc.UserInfo(ctx, key)
+	return ms.svc.UserInfo(ctx, token)
+}
+
+func (ms *metricsMiddleware) UpdateMetadata(ctx context.Context, token string, metadata map[string]interface{}) (email string, err error) {
+	defer func(begin time.Time) {
+		ms.counter.With("method", "user_update").Add(1)
```
Comment (anovakovic01) [inline users/api/metrics.go]: Same here.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: users/api/http/endpoint.go [hunk 8]
```diff
@@ -90,6 +90,23 @@ func userInfoEndpoint(svc users.Service) endpoint.Endpoint {
 	}
 }
 
+func userUpdateMetaEndpoint(svc users.Service) endpoint.Endpoint {
```
Comment (anovakovic01) [inline users/api/http/endpoint.go]: Please rename this method name.

File: users/api/http/requests.go [hunk 9]
```diff
@@ -28,6 +28,18 @@ func (req viewUserInfoReq) validate() error {
 	return nil
 }
 
+type updateUserReq struct {
+	Token string
+	User  users.User
```
Comment (anovakovic01) [inline users/api/http/requests.go]: First, I want to ask why is `Token` public? Second, I still think that here you should have fields that you're updating (for now just `metadata`). If you leave it this way, I can pass email and pass, and the request will be accepted.

File: users/api/logging.go [hunk 10]
```diff
@@ -74,7 +74,20 @@ func (lm *loggingMiddleware) UserInfo(ctx context.Context, key string) (u users.
 		lm.logger.Info(fmt.Sprintf("%s without errors.", message))
 	}(time.Now())
 
-	return lm.svc.UserInfo(ctx, key)
+	return lm.svc.UserInfo(ctx, token)
+}
+
+func (lm *loggingMiddleware) UpdateUser(ctx context.Context, token string, u users.User) (err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method update_user_info for user %s took %s to complete", u.Email, time.Since(begin))
```
Comment (anovakovic01) [inline users/api/logging.go]: Change to `update_user`

File: users/api/metrics.go [hunk 11]
```diff
@@ -56,13 +56,22 @@ func (ms *metricsMiddleware) Identify(key string) (string, error) {
...
 		ms.latency.With("method", "user_info").Observe(time.Since(begin).Seconds())
 	}(time.Now())
 
-	return ms.svc.UserInfo(ctx, key)
+	return ms.svc.UserInfo(ctx, token)
+}
+
+func (ms *metricsMiddleware) UpdateUser(ctx context.Context, token string, u users.User) (err error) {
+	defer func(begin time.Time) {
+		ms.counter.With("method", "update_user_info").Add(1)
```
Comment (anovakovic01) [inline users/api/metrics.go]: Change to `update_user`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Review by drasko — APPROVED** [review]
LGTM

**Review by drasko — CHANGES_REQUESTED** [review]
Tests need to be added for this new endpoint.

**Comment by manuio** [comment]
Add missing tests in this issue: https://github.com/mainflux/mainflux/issues/926

**Review by drasko — APPROVED** [review]

**Pull request merged**